### ISSUE TITLE
Allow for extendibility of Roole

### ISSUE THREAD

**Issue by iamcarrico** [curvedmark/roole#67]
One of the pieces that has led to the success of Sass / Compass is that Compass has the ability of making extensions to have common functionality be put into community projects. Good examples of this are susy, breakpoint, respond-to, compass-normalize, etc.

Here are some ground pieces that should help:
- We should allow these extensions to be downloaded with npm
- If we need a framework to load all of these pieces with a settings file of some sort, [Grunt.js](gruntjs.com) should be considered. 
- This will allow for functions such as the ones in #15 and #46 to be in JavaScript and possibly additions outside of Roole. 

I am still going through the code base, and examining pieces--- so parts of this might be enabled already. But I figured the issue should be brought up to talk about the "how".

**Comment by curvedmark** [comment]
I will think about this one after v0.4.0.

**Comment by iamcarrico** [comment]
Rodger--- what can I do to help? I have been going through the source code to ensue I know how everything is working. Although I am still going through it, I can be available to help push the development along.

**Comment by curvedmark** [comment]
Thanks for wanting to help. :)

But at this stage, I guess no much code could be written for this feature, only some plannings.

I'm still considering what extensions should be allowed to do. I know adding built-in function should definitely be on the list, but anything else? Should extensions be allowed to temper with vender prefixing or even adding new `@something` constructs?

Since I haven't finished #15 yet, I I'm not sure how built-in functions should be implement, let alone designing an API to add them. So the API should wait until 0.4.0.

I like your idea about installing extensions with npm. To load them, a command line flag should be enough (users specify which extensions they want to use, and Roole will `require()` them), no need to use a setting file or even grunt.js. But that exposes some problems for browsers, since you can not directly put code written for nodejs in browsers.

The core uses `build/commonjs-stripper` to convert the code for browser use, maybe we can expose it to extension authors, asking them to release two files, one for nodejs, one for browsers, but that sounds a bit cumbersome.

I should be able to land #15 in a day or two, and probably with a few built-in functions implemented. Then we can think about this feature more seriously.

**Comment by iamcarrico** [comment]
So, I need to take a better look into #15... BUT here are some thoughts for yah. And again, I can look into doing part of this now that milestone 0.4.0 is seemingly hit. 

An extension should be able to provide two things:
1) Its own .roo files being added in. So in my project I can do a @include('my_extension_name); and that will include my extensions .roo file, and with it functions, variables, etc, that I use. For example, I can create a extension like Susy for compass, and have the large amount of functions to help make a fluid design. 

2) Ability to add in .js to the system. I think this should primarily be in the form of adding in functions, that express themselves via js and not within a .roo file. We could have this be a different namespace, but it would require that the js gets compiled into the client side. This is part of what I mean when I say we should have grunt.js / settings file. Because in there, specific extensions could be added in, then when a `make roole` is ran, it could compile that js within the main roole.js file. I don't think that will be easy, but I think it will be necessary. I definitely do NOT think that this should be able to add in extra grammar, types of vars, etc. Anything of that nature should be PR'd into the main roole repo for all to use (if it becomes necessary).

**Comment by curvedmark** [comment]
1. If it's just a .roo file, then you should be able to just `@import` the file, and variables inside it will be exposed.
2. What do you mean by "different namespace"?

**Comment by iamcarrico** [comment]
1) So, I don't mean a simple @import, and the .roo file gets added. At least not directly. I want to be able to within my site files have a package.json file that will include a community-driven project. Within that, I want to have a .roo file that I can @import into the project. Thereby, we can have extensions of code that are reusable, and updatable through npm.

2) Well, currently functions are saved as variables. How can we have functions be completely within the JavaScript, but be able to use them within a .roo file? I will try and come up with some ideas this afternoon.

**Comment by curvedmark** [comment]
1. We can make `@import` work like `require()`. If the importing path is, for example `@import 'tabs';`, which doesn't start with `/`, `./` or `../`, Roole will try to load the `.roo` file from `node_modules/tabs` folder (like node, it should recursively try parent folders if the `node_modules` folder couldn't be found in the current folder).
   
   The `tabs` module can specifiy a `package.json` file, which can contain a  `roole` entry. This entry tells Roole which `.roo` file to load, for example `{"roole": "path/to/file.roo"}`.
2. We can expose something like `roole.bif` that you can add methods to:
   
   ```
   var roole = require('roole');
   var Node = roole.Node;
   
   roole.bif.len = function(node) {
       var number = node.type !== 'list' ? 1 : node.children.length;
       return Node('number', [number], {loc: this.loc});
   };
   ```
   
   store this file in `node_modules/len` (along with a `package.json`) for example , and run `roole -u len` (which will `require()` the `len` module), then a `$len()` function will exist when compiling:
   
   ```
   body {
       -foo: $len(a b); // 2
   }
   ```
   
   Two things to notice: `roole.Node` is exposed, `this` inside the function refers to the function call node.
   
   This is still a rough idea, though.

Both types of extensions expose some problems for browsers.
1. For it to work, browsers should first download the `package.json`, find the `roole` entry, and then download that file, doesn't sound very optimal. One solution might be that users are required to link the extension like this:
   
   ```
   <link rel="stylesheet/roole" src="path/to/file.roo" data-roole="tabs">
   ```
   
   telling Roole that this file is a 'tabs' module, and `@import 'tabs'` should import this file.
2. For it to work, I'm thinking of using [component](https://github.com/component/component), still evaluating it.

**Comment by iamcarrico** [comment]
1) Perfect, methinks. There is always the option as well of having a package.json within the main directory, and have it have some settings:

``` javascript
{
  "name": "Name",
  "description": "This is my settings for this stuff",
  "author": "Me <name@email.com>",
  "dependencies": {
    "roole": ">= 0.5.0",
    "tabs": ">= 0.2.0"
  },
  "roole": {
    "directories": {
      "roole": "roo",
      "css": "cdd",
      "images": "img",
      "javascript": "js"
    },
    "requires": {
      "tabs": ">= 0.2.0"
    }
  }
}
```

Then when you do a make, roole will automatically grab those files and put them into a /roo/roole_modules/ file, where they can be used on the front end. This can be used for both the js and the .roo files that an extension would bring in.

**Comment by iamcarrico** [comment]
So--- here is what I am going to do. Since I think you have a better idea of what is in your head as far as making roole goes--- I am going to start making an extension for it to do a lot of styling work I am used to doing. I will make it to your specifications here, and it will provide a good use-case senario to test against.

**Comment by curvedmark** [comment]
Why does roole need info like:

```
"directories": {
  "roole": "roo",
  "css": "cdd",
  "images": "img",
  "javascript": "js"
}
```

and what `requires` is for? since you already listed it in the root `dependencies`.

**Comment by iamcarrico** [comment]
So the first bit was an example of settings that roole might be able to use. Again, pulling from compass's ability to have functions like `image-url('thumbs/image.gif');` that will ensure the proper URL is written when it is compiled. Having a roole and css directory can also give roole the ability of just selecting all .roo files within that directory, and compiling them to the css directory. Making it easier to use. 

The second, may or may not be necessary... Is there a time where a theme of some sort would need another node module that wouldn't have to be imported into roole. We could just ignore that by ignoring anything pulled in that doesnt have specific roole settings. I will leave that one to you.

**Comment by curvedmark** [comment]
`@import` path resolution is is implemented in master, see #100 and #101. As for the url, see #99 for the explanation  of `$__dirname`, which should cover most cases of external resource path resolution.
